plugins/@grouparoo/pipedrive/src/lib/export/destinationMappingOptions.ts

```diff
@@ -0,0 +1,172 @@
+    address: "string",
+    daterange: "date",
+
+    // TODO
+    time: null, // time in the format of "00:00:00"
+    timerange: null,
+    set: null, // multiple options
+  };
+
+  const grouparooType = map[pipedriveType];
+  if (grouparooType) {
+    return grouparooType;
+  }
+
+  return null;
+};
+
+type KnownPersonField = {
+  key: string;
+  pipedriveKey: string;
+  type: DestinationMappingOptionsResponseTypes;
+  required: boolean;
+  important: boolean;
+  groupMembershipField: boolean;
+};
+
+export const getKnownPersonFieldMap = async (
+  client: any,
+  cacheData: PipedriveCacheData,
+  update: boolean = false
+): Promise<{ [fieldName: string]: string }> => {
+  const known = await getKnownPersonFields(client, cacheData, update);
+  const out = {};
+  for (const field of known) {
+    out[field.key] = field.pipedriveKey;
+  }
+  return out;
+};
+
+const getKnownPersonFields = async (
+  client: any,
+  cacheData: PipedriveCacheData,
+  update: boolean = false
+): Promise<KnownPersonField[]> => {
+  const { appId, appOptions } = cacheData;
+  const cacheDurationMs = 1000 * 60 * 10; // 10 minutes
+  const cacheKey = ["getKnownPersonFields", appOptions];
+  const read = !update; // if updating, skip the read from cache. still write.
+  return objectCache(
+    { objectId: appId, cacheKey, cacheDurationMs, read },
+    async () => {
+      return fetchKnownPersonFields(client);
+    }
+  );
+};
+
+export const fetchKnownPersonFields = async (
+  client: any
+): Promise<KnownPersonField[]> => {
+  const required = ["name", "email"];
+  const allowedFields = [...required, "phone", "label"];
+
+  // Allow setting "End of" for date ranges and "Currency" for monetary fields
+  // The keys for these are like cd77a9b30948242ef1ef57acbe28c0aa054b8fde_until
+  const subfieldsRegex = /(_until|_currency)$/;
+
+  const {
+    data: fields,
+  } = await client.PersonFieldsController.getAllPersonFields();
+
+  const out: KnownPersonField[] = [];
+
+  for (const field of fields) {
+    // Some fields are system fields that don't make sense to manually update (counts, computed fields, auto timestamps...)
+    // There's no explicit flag for this, but we check for the ones whose field definition can't be modified.
+    // Some of those fields should be available though, so we let them through.
+    if (!field.edit_flag && !allowedFields.includes(field.key)) continue;
+
+    // Most subfields can not be edited, but some can be
+    if (field.is_subfield && !field.key.match(subfieldsRegex)) continue;
+
+    const type = mapTypesToGrouparoo(field.field_type, field.name);
+    if (type) {
+      out.push({
+        key: field.name,
+        pipedriveKey: field.key,
+        required: required.includes(field.key),
+        important: field.important_flag, // they have a built-in (paid) feature to mark fields as important
+        groupMembershipField: field.name.startsWith(GROUP_FIELD_PREFIX), // used to hide our fields from destinationMappingOptions
```

Comment:
That's a nice helper to simplify the UI (don't show our custom group fields)

Reply:
Good catch! Yeah, the pipedrive library uses a singleton pattern. Hadn't considered the possibility of having two different apps. I tried cloning the pipedrive object in multiple ways, but it doesn't work since internally it works under the assumption that it is a singleton. The library didn't do much over the standard REST APIs so I implemented our own client.